compiler/Main.hs

```diff
@@ -72,14 +74,17 @@ compile (file:files) = runGen $ do
               infered <- inferProgram env desugared
               case infered of
                 Right (prog, env') ->
-                  let (var, tys) = extractToplevels parsed
-                      (var', tys') = extractToplevels resolved
-                  in pure $ Right (tops ++ prog
-                                  , scope { RS.varScope = RS.insertN (RS.varScope scope) (zip var var')
-                                          , RS.tyScope  = RS.insertN (RS.tyScope scope)  (zip tys tys')
-                                          }
-                                  , modScope'
-                                  , env')
+                  case runVerify (verifyProgram prog) of
```

Comment:
I wonder if we should look into shifting this into a monad error or something.

Reply:
We probably should, but that'll be Hard tm

Reply:
because they all have different error types

Reply:
I was thinking something like `` `catchError` (throwError . Verify)`, where `Verify` is some constructor for a union of all error types.

Or we do something with existentials :p.

Reply:
That won't work.